95.md

```diff
@@ -0,0 +1,108 @@
+NIP-95
+======
+
+Relay File Storage
+------------------
+
+`draft` `optional`
+
+This NIP defines a way to store binary data on relays. It supports parallel uploads and resumable downloads.
+
+## Upload
+
+To upload a file, first client must convert its bytes to base64. It may do it in chunks made of multiples of 3 bytes or in one go.
+The base64 data will be used on the `.content` field of one or more `kind:1064` events to be uploaded to a relay.
+
+The chunk events may be sent on separate websockets connections for parallel uploads.
+
+## Event
+
+An event of `kind:1064` represents a file chunk. A client may choose to split a file
+into multiple chunks or use a single event.
+
+The event `.pubkey` is NOT the uploader's one but a randomly generated pubkey. All chunks
+use the same pubkey. The pubkey MUST NOT be reused on events other than the
+`kind:1064` event set of a single file.
+**The pubkey is what is used to uniquely identify the file made of a set of chunk events**
+
+The event `.content` is the base64-encoded file data.
+
+It has a `c` tag. Its first falue is the chunk number while the second one is the total number of chunks.
+The client can use the `c` tag to ask for missing chunks when resuming a download.
+
+Example:
+
+```js
+{
+  "id": "<id>",
+  "pubkey": "<same-random-pubkey-for-all-event-chunks>"
+  "kind": 1064,
+  "tags": [
+    ["c", "5", "20"] // required; this is the 5th chunk of a total of 20
+  ],
+  "content": "<base64-chunk>",
+  "created_at": <timestamp>,
+  "sig": "<signature>"
+}
+```
+
+## Metadata
+
+Metadata should be added using [NIP-94](94.md) or other metadata-related NIP.
+
+## File Relay
+
+Client should upload to user's "file relays", which use the [NIP-65](65.md) `f` flag.
+When downloading a file uploaded with this NIP, it should search on the uploader's "file relays".
+
+**Relays must NOT honor `kind:5` deletion events referencing file chunk events.** Deletion
```

Comment:
No need for this. I might want to get a separate chunk of the file in each relay. Deleting chunks should be possible and it should be fine.

Reply:
I thought of the following flow, maybe it is bogus:
- An userA uploads a file.
- Another userB sees it on his client and asks client to copy and upload it too (to make it his file) and both users happen to use the same file relay.
- UserB client instead of re-uploading it, just registers userB as an owner/uploader of the file that is already on the relay.

Now that userA and userB both own the file, we can't let userA delete the chunks.. userA can just unregister himself as not an uploader/owner anymore. The file is deleted if there is no registered uploader.

Reply:
Are you trying to register many owners for shared chunks and only allow deletion when all owners request or stop using the chunk?

Reply:
Yes, anyone the file relay authorizes can become an owner ("uploader) of a file chunk set.

There is an "uploader event" for that. When no "uploader event" is present on the relay anymore (deleted with `kind:5`), the file relay is free to automatically delete the chunk set (not using `kind:5` here, just auto-deletes).